src/algebra/ring/projections.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2022 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import algebra.ring.basic
+
+/-!
+# Idempotents
+
+This file defines idempotents for an arbitary multiplication and proves some basic results:
+
+* In a semigroup, the product of two commuting idempotents is an idempotent;
+* In a (non-associative) ring, p is an idempotent if and only if 1-p is an idempotent.
+
+## Tags
+
+projection, idempotent
+-/
+
+variables {M : Type*} [has_mul M]
+
+/--
+An element `p` is said to be idempotent if `p * p = p`
+-/
+def is_idempotent_elem (p : M) : Prop := p*p = p
+
+lemma idempotent_elem_def {p : M} (h : is_idempotent_elem p) : p*p = p := h
+
+namespace is_idempotent_elem
+
```

Comment:
You should probably add the lemmas that `0` and `1` are idempotent!

Reply:
Done. I've also shown that `0ᶜ = 1` and `1ᶜ = 0`.